### ISSUE TITLE
 Set TextField's cursor position

### ISSUE THREAD

**Issue by fabiomicheldossantos** [flutter/flutter#94058]
Good afternoon. I have a problem that I've been trying to solve for days and I can't. I have the following code where the TextField occupies the entire body of the Scaffold and starts in read-only mode. After a double click, the readOnly state is changed to false for user editing. The problem is that I want the TextField cursor to be positioned exactly where the double click occurred, but it doesn't. This "problem" started after a recent Flutter update. Does anyone know how to fix this?

```
class _HomeViewState extends State<HomeView> {
  TextEditingController contentController = TextEditingController();
  bool visualization = true;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: LayoutBuilder(
        builder: (context, constraint) {
          return SingleChildScrollView(
            padding: EdgeInsets.symmetric(horizontal: 16, vertical: 10),
            child: ConstrainedBox(
              constraints: BoxConstraints(
                minHeight: (constraint.maxHeight - 20),
              ),
              child: IntrinsicHeight(
                child: GestureDetector(
                  onDoubleTap: () {
                    setState(() {
                      visualization = !visualization;
                    });
                  },
                  child: TextField(
                    controller: contentController,
                    decoration: InputDecoration(
                      border: InputBorder.none,
                      contentPadding: EdgeInsets.zero,
                    ),
                    expands: true,
                    keyboardType: TextInputType.multiline,
                    maxLines: null,
                    minLines: null,
                    readOnly: visualization ? true : false,
                    style: TextStyle(fontSize: 22),
                  ),
                ),
              ),
            ),
          );
        },
      ),
    );
  }
}
```

**Comment by nocsg** [comment]
hi @fabiomicheldossantos ! Because the code `bool visualization = true;`, the value of visualization is always `true` whenever you reset state of page. you can modify your code like  `visualization?? true`.

**Comment by mamuseferha** [comment]
Hi @fabiomicheldossantos I believe this is the expected behaviour when the initial text of the field is empty. Can you confirm if the behaviour persists when you set initial text in the controller? 

 ```
TextEditingController contentController =
      TextEditingController(text: 'Hi There');
```

**Comment by fabiomicheldossantos** [comment]
> hi @fabiomicheldossantos ! Because the code `bool visualization = true;`, the value of visualization is always `true` whenever you reset state of page. you can modify your code like `visualization?? true`.

**Comment by fabiomicheldossantos** [comment]
> hi @fabiomicheldossantos ! Because the code `bool visualization = true;`, the value of visualization is always `true` whenever you reset state of page. you can modify your code like `visualization?? true`.

Thanks for your reply but when I double click the screen the state is changed to false as expected. =)

**Comment by fabiomicheldossantos** [comment]
> Hi @fabiomicheldossantos I believe this is the expected behaviour when the initial text of the field is empty. Can you confirm if the behaviour persists when you set initial text in the controller?
> 
> ```
> TextEditingController contentController =
>      TextEditingController(text: 'Hi There');
> ```

Hi, @mamuseferha. Sorry for not being more specific, but the problem is when there are several lines written. Example: if there are ten lines written and the user double-clicks in the middle of the fifth line, it is expected that the cursor would be positioned there for editing.

**Comment by mamuseferha** [comment]
Hi @fabiomicheldossantos, I tested this on the latest stable and master channel using the sample code below and it is not reproducible. When I double click on any spot in the textfield the cursor is positioned there.

What device does it occur, and also can you provide your `flutter doctor -v`?

<details>
<summary>code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: MyWidget(),
        ),
      ),
    );
  }
}

class MyWidget extends StatefulWidget {
  const MyWidget({Key? key}) : super(key: key);

  @override
  State<MyWidget> createState() => _MyWidgetState();
}

class _MyWidgetState extends State<MyWidget> {
  TextEditingController contentController = TextEditingController(
      text:
          "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.");
  bool visualization = true;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: LayoutBuilder(
        builder: (context, constraint) {
          return SingleChildScrollView(
            padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 10),
            child: ConstrainedBox(
              constraints: BoxConstraints(
                minHeight: (constraint.maxHeight - 20),
              ),
              child: IntrinsicHeight(
                child: GestureDetector(
                  onDoubleTap: () {
                    setState(() {
                      visualization = !visualization;
                    });
                  },
                  child: TextField(
                    controller: contentController,
                    decoration: const InputDecoration(
                      border: InputBorder.none,
                      contentPadding: EdgeInsets.zero,
                    ),
                    expands: true,
                    keyboardType: TextInputType.multiline,
                    maxLines: null,
                    minLines: null,
                    readOnly: visualization ? true : false,
                    style: const TextStyle(fontSize: 22),
                  ),
                ),
              ),
            ),
          );
        },
      ),
    );
  }
}

```
</details>

<details>
<summary>flutter doctor -v</summary>

```bash
[✓] Flutter (Channel master, 2.6.0-12.0.pre.826, on macOS 11.6.1 20G224 darwin-x64, locale en-US)
    • Flutter version 2.6.0-12.0.pre.826 at /Users/flo/fvm/versions/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 6c1455a23a (5 hours ago), 2021-11-23 21:13:05 -0500
    • Engine revision 539239e256
    • Dart version 2.16.0 (build 2.16.0-42.0.dev)
    • DevTools version 2.8.0

[!] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/flo/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 13.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] VS Code (version 1.62.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.28.0

[✓] Connected device (5 available)
    • SM G975U (mobile)                  • R58M95MA95E                          • android-arm64  • Android 11 (API 30)
    • Android SDK built for x86 (mobile) • emulator-5554                        • android-x86    • Android 11 (API 30) (emulator)
    • iPhone 13 (mobile)                 • EF1251D3-C9C4-475E-9F91-58CD010FDA48 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-0 (simulator)
    • macOS (desktop)                    • macos                                • darwin-x64     • macOS 11.6.1 20G224 darwin-x64
    • Chrome (web)                       • chrome                               • web-javascript • Google Chrome 93.0.4577.63

! Doctor found issues in 1 category.
```
```bash
[✓] Flutter (Channel master, 2.6.0-12.0.pre.826, on macOS 11.6.1 20G224 darwin-x64, locale en-US)
    • Flutter version 2.6.0-12.0.pre.826 at /Users/flo/fvm/versions/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 6c1455a23a (5 hours ago), 2021-11-23 21:13:05 -0500
    • Engine revision 539239e256
    • Dart version 2.16.0 (build 2.16.0-42.0.dev)
    • DevTools version 2.8.0

[!] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/flo/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 13.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] VS Code (version 1.62.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.28.0

[✓] Connected device (5 available)
    • SM G975U (mobile)                  • R58M95MA95E                          • android-arm64  • Android 11 (API 30)
    • Android SDK built for x86 (mobile) • emulator-5554                        • android-x86    • Android 11 (API 30) (emulator)
    • iPhone 13 (mobile)                 • EF1251D3-C9C4-475E-9F91-58CD010FDA48 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-0 (simulator)
    • macOS (desktop)                    • macos                                • darwin-x64     • macOS 11.6.1 20G224 darwin-x64
    • Chrome (web)                       • chrome                               • web-javascript • Google Chrome 93.0.4577.63

! Doctor found issues in 1 category.
```

</details>

Thank you

**Comment by fabiomicheldossantos** [comment]
Hi @mamuseferha, I'm using a Samsung Galaxy A9, 2016, but in the emulator with Nexus 5X API 29 the problem also occurs.  

Follow the ``flutter doctor -v``.

````
[√] Flutter (Channel stable, 2.5.3, on Microsoft Windows [versÃ£o 10.0.19044.1348], locale pt-BR)
    • Flutter version 2.5.3 at C:\src\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 18116933e7 (6 weeks ago), 2021-10-15 10:46:35 -0700
    • Engine revision d3ea636dc5
    • Dart version 2.14.4

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\FMS\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Android Studio (version 2020.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[√] VS Code (version 1.56.2)
    • VS Code at C:\Users\FMS\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension can be installed from:
       https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[√] Connected device (3 available)
    • SM A910F (mobile) • 6d2c5e03 • android-arm64  • Android 8.0.0 (API 26)
    • Chrome (web)      • chrome   • web-javascript • Google Chrome 96.0.4664.45
    • Edge (web)        • edge     • web-javascript • Microsoft Edge 96.0.1054.29
    ! Device emulator-5554 is offline.

• No issues found!
````

Thanks for your attention. Good night!

**Comment by mamuseferha** [comment]
Hi @fabiomicheldossantos, I tested this on the latest stable and master channel using the following devices but it is still not reproducible? 

Nexus 5X API 29 ✅
Nexus 5X API 26 ✅
Nexus 5X API 31 ✅
Samsung Galaxy S10+ API 30 ✅

I'll be leaving this open and labelling for further investigation as I cannot reproduce on my end.

<details>
<summary>output</summary>

https://user-images.githubusercontent.com/92717368/143430434-889eba5e-4d90-4bc6-ae10-a6ff0a101e0c.MP4


</details>


<details>
<summary>flutter doctor -v</summary>

```bash
[✓] Flutter (Channel stable, 2.5.3, on macOS 11.6.1 20G224 darwin-x64, locale en-US)
    • Flutter version 2.5.3 at /Users/flo/development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 18116933e7 (6 weeks ago), 2021-10-15 10:46:35 -0700
    • Engine revision d3ea636dc5
    • Dart version 2.14.4

[!] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/flo/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 13.1, Build version 13A1030d
    • CocoaPods version 1.11.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] VS Code (version 1.62.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.28.0

[✓] Connected device (6 available)
    • SM G975U (mobile)                  • R58M95MA95E                          • android-arm64  • Android 11 (API 30)
    • Android SDK built for x86 (mobile) • emulator-5554                        • android-x86    • Android 10 (API 29) (emulator)
    • Android SDK built for x86 (mobile) • emulator-5556                        • android-x86    • Android 8.0.0 (API 26)
      (emulator)
    • iPhone 13 (mobile)                 • EF1251D3-C9C4-475E-9F91-58CD010FDA48 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-15-0 (simulator)
    • macOS (desktop)                    • macos                                • darwin-x64     • macOS 11.6.1 20G224 darwin-x64
    • Chrome (web)                       • chrome                               • web-javascript • Google Chrome 93.0.4577.63

! Doctor found issues in 1 category.
```
```bash
[✓] Flutter (Channel master, 2.6.0-12.0.pre.837, on macOS 11.6.1 20G224 darwin-x64, locale en-US)
    • Flutter version 2.6.0-12.0.pre.837 at /Users/flo/fvm/versions/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 341e771b12 (8 hours ago), 2021-11-24 19:33:04 -0500
    • Engine revision e99aba6a38
    • Dart version 2.16.0 (build 2.16.0-46.0.dev)
    • DevTools version 2.8.0

[!] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/flo/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 13.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] VS Code (version 1.62.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.28.0

[✓] Connected device (6 available)
    • SM G975U (mobile)                  • R58M95MA95E                          • android-arm64  • Android 11 (API 30)
    • Android SDK built for x86 (mobile) • emulator-5554                        • android-x86    • Android 10 (API 29) (emulator)
    • Android SDK built for x86 (mobile) • emulator-5556                        • android-x86    • Android 8.0.0 (API 26) (emulator)
    • iPhone 13 (mobile)                 • EF1251D3-C9C4-475E-9F91-58CD010FDA48 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-0 (simulator)
    • macOS (desktop)                    • macos                                • darwin-x64     • macOS 11.6.1 20G224 darwin-x64
    • Chrome (web)                       • chrome                               • web-javascript • Google Chrome 93.0.4577.63

! Doctor found issues in 1 category.
```

</details>

Thank you

**Comment by fabiomicheldossantos** [comment]
Ok, @mamuseferha. Thank you for all your efforts.

**Comment by fabiomicheldossantos** [comment]
Studying the code a bit more, I found that until Flutter version 2.2.3, the desired behavior was working. Whenever the user double-clicked, the mouse cursor was positioned at the click location (considering the existence of multiline text) and the text field was enabled for editing, that is, if the user clicked halfway through the fifth line, the cursor was positioned there, and so on. After updating to version 2.5.3 this cursor behavior stopped working so that when the user double-clicks on it, it always appears positioned in the last place where it was already located. I'm still new to Flutter and the programming world, but after a long research, I found that the change in this behavior was due to the alteration of five Flutter files, namely:

     packages/flutter/lib/src/cupertino/desktop_text_selection.dart;
     packages/flutter/lib/src/cupertino/text_selection.dart;
     packages/flutter/lib/src/material/desktop_text_selection.dart;
     packages/flutter/lib/src/material/text_selection.dart;
     packages/flutter/lib/src/widgets/text_selection.dart.
